File: public/scripts/generate_report.js
#!/usr/bin/env node

/**
 * Cloud Storage Optimizer - Storage Report Generator (Node.js)
 * ==============================================================
 * This script scans a local directory and generates a JSON report
 * that can be uploaded to the Cloud Storage Optimizer web app.
 * 
 * Usage:
 *     node generate_report.js
 * 
 * Requirements:
 *     - Node.js 14+
 *     - No external dependencies (uses only built-in modules)
 * 
 * Created for: https://your-app.com
 */

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const readline = require('readline');

// =======================================
// Configuration
// =======================================
const CHUNK_SIZE = 8192; // 8KB chunks for file reading
const MAX_FILE_SIZE_FOR_HASH = 100 * 1024 * 1024; // 100MB

// =======================================
// Helper Functions
// =======================================

/**
 * Calculate MD5 hash of a file
 */
function getFileHash(filepath) {
  try {
    const stats = fs.statSync(filepath);
    
    // Skip hashing for very large files
    if (stats.size > MAX_FILE_SIZE_FOR_HASH) {
      return `large_file_${stats.size}`;
    }
    
    const hash = crypto.createHash('md5');
    const data = fs.readFileSync(filepath);
    hash.update(data);
    return hash.digest('hex');
  } catch (error) {
    return `error_${error.message.substring(0, 20)}`;
  }
}

/**
 * Determine file type based on extension
 */
function getFileType(filename) {
  const ext = path.extname(filename).toLowerCase();
  
  const typeMap = {
    // Images
    '.jpg': 'image', '.jpeg': 'image', '.png': 'image',
    '.gif': 'image', '.bmp': 'image', '.svg': 'image',
    '.heic': 'image', '.webp': 'image',
    
    // Videos
    '.mp4': 'video', '.avi': 'video', '.mov': 'video',
    '.mkv': 'video', '.flv': 'video', '.wmv': 'video',
    '.webm': 'video', '.m4v': 'video',
    
    // Documents
    '.pdf': 'document', '.docx': 'document', '.doc': 'document',
    '.txt': 'document', '.rtf': 'document', '.odt': 'document',
    '.xlsx': 'document', '.xls': 'document', '.pptx': 'document',
    
    // Archives/Backups
    '.zip': 'backup', '.tar': 'backup', '.gz': 'backup',
    '.rar': 'backup', '.7z': 'backup', '.bz2': 'backup',
    
    // Logs
    '.csv': 'log', '.log': 'log', '.json': 'log',
    
    // Code
    '.js': 'code', '.py': 'code', '.html': 'code',
    '.css': 'code', '.java': 'code', '.cpp': 'code',
    
    // Audio
    '.mp3': 'audio', '.wav': 'audio', '.flac': 'audio',
    '.m4a': 'audio', '.aac': 'audio'
  };
  
  return typeMap[ext] || 'other';
}

/**
 * Format bytes to human readable format
 */
function formatSize(bytes) {
  const units = ['B', 'KB', 'MB', 'GB', 'TB'];
  let size = bytes;
  let unitIndex = 0;
  
  while (size >= 1024 && unitIndex < units.length - 1) {
    size /= 1024;
    unitIndex++;
  }
  
  return `${size.toFixed(2)} ${units[unitIndex]}`;
}

/**
 * Recursively scan a directory and collect file metadata
 */
function scanDirectory(dirPath, showProgress = true) {
  const files = [];
  let totalSize = 0;
  let fileCount = 0;
  let skipped = 0;
  
  console.log(`\n🔍 Scanning directory: ${dirPath}`);
  console.log('='.repeat(70));
  
  function scanRecursive(currentPath) {
    try {
      const items = fs.readdirSync(currentPath);
      
      for (const item of items) {
        // Skip hidden files/folders
        if (item.startsWith('.')) continue;
        
        const itemPath = path.join(currentPath, item);
        
        try {
          const stats = fs.statSync(itemPath);
          
          if (stats.isDirectory()) {
            // Recursively scan subdirectory
            scanRecursive(itemPath);
          } else if (stats.isFile()) {
            const fileData = {
              name: item,
              size: stats.size,
              lastModified: stats.mtime.toISOString(),
              type: getFileType(item),
              storageClass: 'standard',
              hash: getFileHash(itemPath)
            };
            
            files.push(fileData);
            totalSize += stats.size;
            fileCount++;
            
            // Show progress every 50 files
            if (showProgress && fileCount % 50 === 0) {
              console.log(`  📂 Found ${fileCount.toLocaleString()} files... (${formatSize(totalSize)})`);
            }
          }
        } catch (error) {
          skipped++;
          if (showProgress && skipped <= 5) {
            console.log(`  ⚠️  Skipping ${item}: ${error.message.substring(0, 50)}`);
          }
        }
      }
    } catch (error) {
      console.error(`\n❌ Error scanning ${currentPath}: ${error.message}`);
    }
  }
  
  scanRecursive(dirPath);
  
  if (skipped > 5) {
    console.log(`  ⚠️  ... and ${skipped - 5} more files skipped`);
  }
  
  return { files, totalSize };
}

/**
 * Prompt user for input
 */
function prompt(question) {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });
  
  return new Promise(resolve => {
    rl.question(question, answer => {
      rl.close();
      resolve(answer.trim());
    });
  });
}

// =======================================
// Main Function
// =======================================

async function main() {
  // Print welcome banner
  console.log(`
╔═══════════════════════════════════════════════════════════╗
║                                                           ║
║      Cloud Storage Optimizer                              ║
║      Storage Report Generator (Node.js)                   ║
║                                                           ║
║      Scan your local storage and find optimization        ║
║      opportunities!                                       ║
║                                                           ║
╚═══════════════════════════════════════════════════════════╝
  `);
  
  // Get directory from user
  let directory;
  while (true) {
    console.log('\n📁 Enter the directory path you want to scan:');
    console.log('   Examples:');
    console.log('     Windows: C:\\Users\\YourName\\Documents');
    console.log('     Mac:     /Users/YourName/Documents');
    console.log('     Linux:   /home/yourname/documents');
    console.log();
    
    directory = await prompt('Path: ');
    directory = directory.replace(/['"]/g, ''); // Remove quotes
    
    if (!directory) {
      console.log('❌ Please enter a path');
      continue;
    }
    
    if (!fs.existsSync(directory)) {
      console.log(`❌ Path does not exist: ${directory}`);
      const retry = await prompt('Try again? (y/n): ');
      if (retry.toLowerCase() !== 'y') {
        console.log('Goodbye! 👋');
        process.exit(0);
      }
      continue;
    }
    
    const stats = fs.statSync(directory);
    if (!stats.isDirectory()) {
      console.log(`❌ Path is not a directory: ${directory}`);
      continue;
    }
    
    break;
  }
  
  // Scan the directory
  console.log('\n⏳ Starting scan... (this may take a while for large directories)');
  const { files, totalSize } = scanDirectory(directory, true);
  
  // Print summary
  console.log('\n' + '='.repeat(70));
  console.log('✅ Scan complete!');
  console.log(`   📊 Files found: ${files.length.toLocaleString()}`);
  console.log(`   💾 Total size: ${formatSize(totalSize)}`);
  
  // Ask for output filename
  console.log('\n💾 Save report as:');
  let outputFile = await prompt('   Filename (default: storage_report.json): ');
  if (!outputFile) {
    outputFile = 'storage_report.json';
  }
  if (!outputFile.endsWith('.json')) {
    outputFile += '.json';
  }
  
  // Create report
  const report = {
    files: files,
    metadata: {
      scannedPath: directory,
      scannedAt: new Date().toISOString(),
      totalFiles: files.length,
      totalSize: totalSize
    }
  };
  
  // Save to file
  try {
    fs.writeFileSync(outputFile, JSON.stringify(report, null, 2), 'utf-8');
    
    const fileStats = fs.statSync(outputFile);
    console.log('\n✅ Report saved successfully!');
    console.log(`   📄 File: ${path.resolve(outputFile)}`);
    console.log(`   📊 Size: ${formatSize(fileStats.size)}`);
    
  } catch (error) {
    console.error(`\n❌ Error saving file: ${error.message}`);
    process.exit(1);
  }
  
  // Next steps
  console.log('\n' + '='.repeat(70));
  console.log('🚀 Next Steps:');
  console.log('   1. Go to your Cloud Storage Optimizer web app');
  console.log('   2. Click \'Upload JSON File\'');
  console.log(`   3. Select: ${outputFile}`);
  console.log('   4. Get your optimization recommendations!');
  console.log('\n' + '='.repeat(70));
  console.log('\nThank you for using Cloud Storage Optimizer! 💙\n');
}

// =======================================
// Entry Point
// =======================================

main().catch(error => {
  console.error(`\n❌ Unexpected error: ${error.message}`);
  process.exit(1);
});

File: public/scripts/generate_report.py
#!/usr/bin/env python3
"""
Cloud Storage Optimizer - Storage Report Generator (Python)
==============================================================
This script scans a local directory and generates a JSON report
that can be uploaded to the Cloud Storage Optimizer web app.

Usage:
    python generate_report.py

Requirements:
    - Python 3.6+
    - No external dependencies (uses only built-in modules)

Created for: https://your-app.com
"""

import os
import json
import hashlib
from datetime import datetime
from pathlib import Path
import sys

# =======================================
# Configuration
# =======================================
CHUNK_SIZE = 8192  # For reading files in chunks (8KB)
MAX_FILE_SIZE_FOR_HASH = 100 * 1024 * 1024  # 100MB - don't hash files larger than this

# =======================================
# Helper Functions
# =======================================

def get_file_hash(filepath):
    """
    Calculate MD5 hash of a file.
    For large files (>100MB), we skip hashing to save time.
    """
    try:
        file_size = os.path.getsize(filepath)
        
        # Skip hashing for very large files
        if file_size > MAX_FILE_SIZE_FOR_HASH:
            return f"large_file_{file_size}"
        
        hasher = hashlib.md5()
        with open(filepath, 'rb') as f:
            while chunk := f.read(CHUNK_SIZE):
                hasher.update(chunk)
        return hasher.hexdigest()
    except Exception as e:
        return f"error_{str(e)[:20]}"

def get_file_type(filename):
    """
    Determine file type based on extension.
    """
    ext = Path(filename).suffix.lower()
    
    type_map = {
        # Images
        '.jpg': 'image', '.jpeg': 'image', '.png': 'image', 
        '.gif': 'image', '.bmp': 'image', '.svg': 'image',
        '.heic': 'image', '.webp': 'image',
        
        # Videos
        '.mp4': 'video', '.avi': 'video', '.mov': 'video', 
        '.mkv': 'video', '.flv': 'video', '.wmv': 'video',
        '.webm': 'video', '.m4v': 'video',
        
        # Documents
        '.pdf': 'document', '.docx': 'document', '.doc': 'document',
        '.txt': 'document', '.rtf': 'document', '.odt': 'document',
        '.xlsx': 'document', '.xls': 'document', '.pptx': 'document',
        
        # Archives/Backups
        '.zip': 'backup', '.tar': 'backup', '.gz': 'backup',
        '.rar': 'backup', '.7z': 'backup', '.bz2': 'backup',
        
        # Logs
        '.csv': 'log', '.log': 'log', '.json': 'log',
        
        # Code
        '.py': 'code', '.js': 'code', '.html': 'code',
        '.css': 'code', '.java': 'code', '.cpp': 'code',
        
        # Audio
        '.mp3': 'audio', '.wav': 'audio', '.flac': 'audio',
        '.m4a': 'audio', '.aac': 'audio'
    }
    
    return type_map.get(ext, 'other')

def format_size(bytes_size):
    """Format bytes to human readable format"""
    for unit in ['B', 'KB', 'MB', 'GB', 'TB']:
        if bytes_size < 1024.0:
            return f"{bytes_size:.2f} {unit}"
        bytes_size /= 1024.0
    return f"{bytes_size:.2f} PB"

def scan_directory(path, show_progress=True):
    """
    Scan a directory and collect metadata for all files.
    
    Returns:
        tuple: (list of file data, total size in bytes)
    """
    files = []
    total_size = 0
    file_count = 0
    skipped = 0
    
    print(f"\n🔍 Scanning directory: {path}")
    print("=" * 70)
    
    try:
        for root, dirs, filenames in os.walk(path):
            # Skip hidden directories (starting with .)
            dirs[:] = [d for d in dirs if not d.startswith('.')]
            
            for filename in filenames:
                # Skip hidden files
                if filename.startswith('.'):
                    continue
                
                try:
                    filepath = os.path.join(root, filename)
                    stat = os.stat(filepath)
                    
                    file_data = {
                        'name': filename,
                        'size': stat.st_size,
                        'lastModified': datetime.fromtimestamp(stat.st_mtime).isoformat() + 'Z',
                        'type': get_file_type(filename),
                        'storageClass': 'standard',
                        'hash': get_file_hash(filepath)
                    }
                    
                    files.append(file_data)
                    total_size += stat.st_size
                    file_count += 1
                    
                    # Show progress every 50 files
                    if show_progress and file_count % 50 == 0:
                        print(f"  📂 Found {file_count} files... ({format_size(total_size)})")
                        
                except Exception as e:
                    skipped += 1
                    if show_progress and skipped <= 5:  # Show first 5 errors only
                        print(f"  ⚠️  Skipping {filename}: {str(e)[:50]}")
        
        if skipped > 5:
            print(f"  ⚠️  ... and {skipped - 5} more files skipped")
                    
    except Exception as e:
        print(f"\n❌ Error scanning directory: {e}")
        sys.exit(1)
    
    return files, total_size

# =======================================
# Main Function
# =======================================

def main():
    """Main function - runs the storage report generator"""
    
    # Print welcome banner
    print("""
╔═══════════════════════════════════════════════════════════╗
║                                                           ║
║      Cloud Storage Optimizer                              ║
║      Storage Report Generator (Python)                    ║
║                                                           ║
║      Scan your local storage and find optimization        ║
║      opportunities!                                       ║
║                                                           ║
╚═══════════════════════════════════════════════════════════╝
    """)
    
    # Get directory from user
    while True:
        print("\n📁 Enter the directory path you want to scan:")
        print("   Examples:")
        print("     Windows: C:\\Users\\YourName\\Documents")
        print("     Mac:     /Users/YourName/Documents")
        print("     Linux:   /home/yourname/documents")
        print()
        
        directory = input("Path: ").strip().strip('"').strip("'")
        
        if not directory:
            print("❌ Please enter a path")
            continue
            
        if not os.path.exists(directory):
            print(f"❌ Path does not exist: {directory}")
            retry = input("Try again? (y/n): ").lower()
            if retry != 'y':
                print("Goodbye! 👋")
                sys.exit(0)
            continue
            
        if not os.path.isdir(directory):
            print(f"❌ Path is not a directory: {directory}")
            continue
            
        break
    
    # Scan the directory
    print("\n⏳ Starting scan... (this may take a while for large directories)")
    files_data, total_size = scan_directory(directory, show_progress=True)
    
    # Print summary
    print("\n" + "=" * 70)
    print("✅ Scan complete!")
    print(f"   📊 Files found: {len(files_data):,}")
    print(f"   💾 Total size: {format_size(total_size)}")
    
    # Ask for output filename
    print("\n💾 Save report as:")
    output_file = input("   Filename (default: storage_report.json): ").strip()
    if not output_file:
        output_file = 'storage_report.json'
    if not output_file.endswith('.json'):
        output_file += '.json'
    
    # Create report
    report = {
        'files': files_data,
        'metadata': {
            'scannedPath': directory,
            'scannedAt': datetime.now().isoformat() + 'Z',
            'totalFiles': len(files_data),
            'totalSize': total_size
        }
    }
    
    # Save to file
    try:
        with open(output_file, 'w', encoding='utf-8') as f:
            json.dump(report, f, indent=2, ensure_ascii=False)
        
        file_size = os.path.getsize(output_file)
        print(f"\n✅ Report saved successfully!")
        print(f"   📄 File: {os.path.abspath(output_file)}")
        print(f"   📊 Size: {format_size(file_size)}")
        
    except Exception as e:
        print(f"\n❌ Error saving file: {e}")
        sys.exit(1)
    
    # Next steps
    print("\n" + "=" * 70)
    print("🚀 Next Steps:")
    print(f"   1. Go to your Cloud Storage Optimizer web app")
    print(f"   2. Click 'Upload JSON File'")
    print(f"   3. Select: {output_file}")
    print(f"   4. Get your optimization recommendations!")
    print("\n" + "=" * 70)
    print("\nThank you for using Cloud Storage Optimizer! 💙")
    print()

# =======================================
# Entry Point
# =======================================

if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print("\n\n👋 Cancelled by user. Goodbye!")
        sys.exit(0)
    except Exception as e:
        print(f"\n❌ Unexpected error: {e}")
        sys.exit(1)